### PR TITLE
Reduce ticks

### PR DESCRIPTION
**Pull request by paed01**
Make move to classes to reduce ticks from LazyCompile closures

# Prof result

Iterations: 100000

TL;DR; reduced by approx 66%

## v1

```
 [Summary]:
   ticks  total  nonlib   name
   5583   16.2%   17.0%  JavaScript
  27155   78.9%   82.6%  C++
  18353   53.3%   55.9%  GC
   1546    4.5%          Shared libraries
    121    0.4%          Unaccounted
```

## v2

```
 [Summary]:
   ticks  total  nonlib   name
   3625   30.1%   31.9%  JavaScript
   7672   63.8%   67.6%  C++
   3580   29.8%   31.5%  GC
    684    5.7%          Shared libraries
     52    0.4%          Unaccounted
```

### CODE REVIEW

File: lib/htmlWriter.js [hunk 1]
```diff
@@ -3,18 +3,21 @@
...
-      let markup = "";
-      for (const chunk of chunks) {
-        markup += chunkToMarkup(chunk);
-      }
-      return next(null, markup);
+class HtmlWriter extends Transform {
+  constructor() {
+    super({writableObjectMode: true});
+  }
+  _transform(chunks, encoding, next) {
```
Comment (jonaswalden) [inline lib/htmlWriter.js:10]: why `_transform` and not `transform`?

File: lib/htmlWriter.js [hunk 2]
```diff
@@ -3,18 +3,21 @@
...
     }
-  });
-};
+    return next(null, markup);
+  }
+}
 
+module.exports = function htmlWriter() {
+  return new HtmlWriter();
+};
```
Comment (jonaswalden) [inline lib/htmlWriter.js:23]: why export a function that returns a new instance?

File: lib/transformHtml.js [hunk 3]
```diff
@@ -38,41 +37,20 @@ function asStream(listener) {
...
-      }
-    }
-  };
-  return reader;
+function convert(buf, chunkSize, encoding = "utf8") {
+  return Readable.from(buf, {encoding});
 }
 
-function createESIParser({onopentag, ontext, onclosetag}) {
+function createESIParser(evaluator) {
```
Comment (jonaswalden) [inline lib/transformHtml.js:44]: Not really part of this PR but `ESIParser` class is only used here and it's scoped to this filter / evaluator adapter. Feels weird. Should this belong to the class itself?

**Review by jonaswalden — COMMENTED** [review]
👏 

Also not part of this PR but public API feels a little wonky. Stream interface was crammed in with a bunch of weird exports. Would be nice if it was adjusted to accommodate the streaming interface as if it was the default, and export the old interface as a utility. ..however that would look 🤷 
If we are doing a new major then maybe it's time

File: lib/expression/evaluate.js [hunk 4]
```diff
@@ -3,191 +3,195 @@
...
+  }
+  getFunc(name) {
+    if (!this[name]) throw new Error(`${name} is not implemented`);
+    return this[name];
   }
+}
+
+module.exports = function evaluate(ast, context) {
+  const evaluator = new Evaluator(context);
+  return evaluator.getFunc(ast.type).call(evaluator, ast);
```
[the remark this answers is not in the document]
Comment (paed01) [reply lib/expression/evaluate.js:194]: There are sub-contexts as well, should we keep a cache?

File: lib/htmlWriter.js (hunk 1 above)
Comment (paed01) [reply lib/htmlWriter.js:10]: part of the [api](https://nodejs.org/dist/latest-v14.x/docs/api/stream.html#stream_implementing_a_transform_stream)

File: lib/htmlWriter.js (hunk 2 above)
Comment (paed01) [reply lib/htmlWriter.js:23]: kept to not break the public api, however flaud

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,7 +4,11 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## [Unreleased]
+## [2.0.0] - 2021-12-07
+
+Use classes to reduce ticks
+
+- More descriptive expression error messages
```
[the remark this answers is not in the document]
Comment (paed01) [reply CHANGELOG.md:11]: eslint@8 requires node 12 or later, previously node 9 was required

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,10 +1,19 @@
-# Changelog
+Changelog
+=========
+
 All notable changes to this project will be documented in this file.
 
-The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
-and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
```
Comment (jonaswalden) [inline CHANGELOG.md:5]: don't remove these

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,10 +1,19 @@
...
+Change the public api and refactor to classes
+
+### Breaking
+- The public api has changed, a lot
+- previous default export function is gone, an exported parse function is added instead with a different signature
+- Minimum node version is now 14 since URL behaves' in earlier versions
 
-## [Unreleased]
+### Changes
+- More descriptive expression error messages
```
Comment (jonaswalden) [inline CHANGELOG.md:16]: Use the instructions from [keep a changelog](https://keepachangelog.com/en/1.0.0/) format documentation

```md
### Changed
- Public API
- Minimum node version is now 14 since URL behaves' in earlier versions

### Removed
- Previous default export function is gone, replaced by `parse`

### Fixed
- More descriptive expression error messages
```

**Review by jonaswalden — CHANGES_REQUESTED** [review]

**Review by jonaswalden — APPROVED** [review]
